tests/pyminisketch.py

```diff
@@ -0,0 +1,514 @@
+#!/bin/env python3
+# Copyright (c) 2020 Pieter Wuille
+# Distributed under the MIT software license, see the accompanying
+# file LICENSE or http://www.opensource.org/licenses/mit-license.php.
+
+"""Native Python (slow) reimplementation of libminisketch' algorithms."""
+
+import random
+import unittest
+
+# Irreducible polynomials over GF(2) to use (represented as integers).
+#
+# Most fields can be defined by multiple such polynomials. Minisketch uses the one with the minimal
+# number of nonzero coefficients, and tie-breaking by picking the lexicographically first among
+# those.
+#
+# All polynomials for degrees 2 through 64 (inclusive) are given.
+GF2_MODULI = [
+    None, None,
+    2**2 + 2**1 + 1,
+    2**3 + 2**1 + 1,
+    2**4 + 2**1 + 1,
+    2**5 + 2**2 + 1,
+    2**6 + 2**1 + 1,
+    2**7 + 2**1 + 1,
+    2**8 + 2**4 + 2**3 + 2**1 + 1,
+    2**9 + 2**1 + 1,
+    2**10 + 2**3 + 1,
+    2**11 + 2**2 + 1,
+    2**12 + 2**3 + 1,
+    2**13 + 2**4 + 2**3 + 2**1 + 1,
+    2**14 + 2**5 + 1,
+    2**15 + 2**1 + 1,
+    2**16 + 2**5 + 2**3 + 2**1 + 1,
+    2**17 + 2**3 + 1,
+    2**18 + 2**3 + 1,
+    2**19 + 2**5 + 2**2 + 2**1 + 1,
+    2**20 + 2**3 + 1,
+    2**21 + 2**2 + 1,
+    2**22 + 2**1 + 1,
+    2**23 + 2**5 + 1,
+    2**24 + 2**4 + 2**3 + 2**1 + 1,
+    2**25 + 2**3 + 1,
+    2**26 + 2**4 + 2**3 + 2**1 + 1,
+    2**27 + 2**5 + 2**2 + 2**1 + 1,
+    2**28 + 2**1 + 1,
+    2**29 + 2**2 + 1,
+    2**30 + 2**1 + 1,
+    2**31 + 2**3 + 1,
+    2**32 + 2**7 + 2**3 + 2**2 + 1,
+    2**33 + 2**10 + 1,
+    2**34 + 2**7 + 1,
+    2**35 + 2**2 + 1,
+    2**36 + 2**9 + 1,
+    2**37 + 2**6 + 2**4 + 2**1 + 1,
+    2**38 + 2**6 + 2**5 + 2**1 + 1,
+    2**39 + 2**4 + 1,
+    2**40 + 2**5 + 2**4 + 2**3 + 1,
+    2**41 + 2**3 + 1,
+    2**42 + 2**7 + 1,
+    2**43 + 2**6 + 2**4 + 2**3 + 1,
+    2**44 + 2**5 + 1,
+    2**45 + 2**4 + 2**3 + 2**1 + 1,
+    2**46 + 2**1 + 1,
+    2**47 + 2**5 + 1,
+    2**48 + 2**5 + 2**3 + 2**2 + 1,
+    2**49 + 2**9 + 1,
+    2**50 + 2**4 + 2**3 + 2**2 + 1,
+    2**51 + 2**6 + 2**3 + 2**1 + 1,
+    2**52 + 2**3 + 1,
+    2**53 + 2**6 + 2**2 + 2**1 + 1,
+    2**54 + 2**9 + 1,
+    2**55 + 2**7 + 1,
+    2**56 + 2**7 + 2**4 + 2**2 + 1,
+    2**57 + 2**4 + 1,
+    2**58 + 2**19 + 1,
+    2**59 + 2**7 + 2**4 + 2**2 + 1,
+    2**60 + 2**1 + 1,
+    2**61 + 2**5 + 2**2 + 2**1 + 1,
+    2**62 + 2**29 + 1,
+    2**63 + 2**1 + 1,
+    2**64 + 2**4 + 2**3 + 2**1 + 1
```

Comment:
@glozow Python integers are arbitrary size. You can just do `3**1000000` if you want.

Reply:
How did you verify that they're irreducible?

Reply:
My logic was:
For polynomials over GF(2), possible roots are 0 and 1. If x is a root for polynomial p, p(x) = 0.
So a polynomial p is irreducible when p(0) != 0 and p(1) != 0, i.e. p(0) = p(1) = 1
Basically I plugged in 0 and 1 to all the polynomials and made sure I got 1?

Reply:
That's not actually sufficient. It proves they don't have `(x+0)` or `(x+1)` factors, but they could still have irreducible factors of 2nd degree or higher. For example, `(x^5 + x^4 + 1)` factors into `(x^3 + x + 1)*(x^2 + x + 1)`, but those factors individually are irreducible.

Your reasoning does however help showing why irreducible polynomials over GF(2) need to have a constant term `(1)` (otherwise p(0)=0) and need to have an odd number of nonzero terms (otherwise p(1)=0).

Reply:
Ohhhhh true, makes sense